include/dali/core/mm/malloc_resource.h

```diff
@@ -130,16 +130,63 @@ class pinned_malloc_memory_resource : public pinned_async_resource {
   }
 
   bool do_is_equal(const memory_resource<memory_kind> &other) const noexcept override {
-    return dynamic_cast<const cuda_malloc_memory_resource*>(&other) != nullptr;
+    return dynamic_cast<const pinned_malloc_memory_resource*>(&other) != nullptr;
   }
 
  public:
-  static cuda_malloc_memory_resource &instance() {
-    static cuda_malloc_memory_resource inst;
+  static pinned_malloc_memory_resource &instance() {
+    static pinned_malloc_memory_resource inst;
     return inst;
   }
 };
 
+/**
+ * @brief A memory resource that directly calls cudaMallocManaged and cudaFree.
+ */
+class managed_malloc_memory_resource : public managed_async_resource {
+  const size_t kGuaranteedAlignment = 256;
+
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    if (bytes == 0)
+      return nullptr;
+    if (kGuaranteedAlignment < 256)
+      alignment = 1;  // cudaMallocManaged guarantees suffcient alignment - avoid overhead
+
+    return detail::aligned_alloc([](size_t size) {
+      void *mem = nullptr;
+      CUDA_CALL(cudaMallocManaged(&mem, size | 1));  // |1 to prevent accidental coalescing
```

Comment:
BTW - why `coalescing` is bad for us?

Reply:
Because, as I had learnt the hard way, you can't pass two adjacent buffers allocated with distinct calls to `cudaMalloc` to APIs like `cudaMemcpy` or `cudaMemset` - it will scream, even though the memory is perfectly usable as contiguous and you can do so freely in a kernel.